ADR-0019: GPIO (#686)
home-assistant/architecture: adr/0019-GPIO.md

> # 0018. GPIO

Author: ```suggestion
# 0019. GPIO
```

Date: 2021-12-14

## Status

Accepted

## Context

We currently have a lot of integrations that use or are based on GPIO (including SPI and i2c busses).

- All of these integrations have a low usage count (source Home Assistant Analytics),
  but it adds the same amount of maintenance and review time to the core.
- Most of these integrations are unmaintained.
- The use of GPIO via containers (including our Home Assistant OS, Container
  & Supervised installation methods) is often unstable or complicated.
- Support for GPIO on specific platforms/SBCs: milage varies (e.g., Raspberry Pi on 64 bits platforms).
- The use of these sensors, in general, is better supported and done via dedicated projects like ESPHome.

## Decision

> We no longer accepting integrations that integrate with devices over GPIO.

Reviewer: ```suggestion
We no longer accept integrations that integrate with devices over GPIO.
```

# 0019. GPIO

Date: 2021-12-14

## Status

Accepted

## Context

We currently have a lot of integrations that use or are based on GPIO (including SPI and i2c busses).

- All of these integrations have a low usage count (source Home Assistant Analytics),

>   but it adds the same amount of maintenance and review time to the core.

Reviewer 2: ```suggestion
  but add the same amount of maintenance and review time to the core.
```

Author: `it adds` or `adds`, but just `add` is incorrect in this case :)

Reviewer 2: Take out the part before the comma and you have "All of these integrations ... add the same amount of ..."

I kinda see the "it" but to me it's confusing as the sentence started with a plural subject but seemed to flipped to singular. Alternatively, something like "...to support these integrations"  on the end of the sentence works and looking at it now I think that's the point?

Author: You are right ☺️ 
Thanks!

- Most of these integrations are unmaintained.
- The use of GPIO via containers (including our Home Assistant OS, Container
  & Supervised installation methods) is often unstable or complicated.

> - Support for GPIO on specific platforms/SBCs: milage varies (e.g., Raspberry Pi on 64 bits platforms).

Reviewer 3: ```suggestion
- Support for GPIO on specific platforms/SBCs: mileage varies (e.g., Raspberry Pi on 64 bits platforms).
```

Author: Thanks! 👍

- The use of these sensors, in general, is better supported and done via dedicated projects like ESPHome.

## Decision

- We no longer accept integrations that integrate with devices over GPIO.
- We identify, deprecate and remove integrations that use GPIO.
- It will still be possible to have custom integrations that use GPIO.

> 

Reviewer 2: Does this need a comment on whether HAOS will continue to provide and support a means of accessing GPIO in the long term

(Not sure if this applies but I assume there is some effort in HAOS to provide reliable pass through of those pins)

Author: I dunno, I've kept that out of scope for this ADR (which is focused solely on the integrations parts). I don't see a reason to remove support from that 🤷

Date: 2021-12-14

## Status

Accepted

## Context

We currently have a lot of integrations that use or are based on GPIO (including SPI and i2c busses).

- All of these integrations have a low usage count (source Home Assistant Analytics),
  but it adds the same amount of maintenance and review time to the core.
- Most of these integrations are unmaintained.
- The use of GPIO via containers (including our Home Assistant OS, Container
  & Supervised installation methods) is often unstable or complicated.
- Support for GPIO on specific platforms/SBCs: mileage varies (e.g., Raspberry Pi on 64 bits platforms).
- The use of these sensors, in general, is better supported and done via dedicated projects like ESPHome.

## Decision

- We no longer accept integrations that integrate with devices over GPIO.
- We identify, deprecate and remove integrations that use GPIO.
- It will still be possible to have custom integrations that use GPIO.

GPIO, in this case, means interfacing directly with the individual GPIO pins
(or I/O lines) on the board running Home Assistant. This includes bus
communications like I2C and SPI.

Devices connected on GPIO that are exposed on the system as regular serial

> devices, are exempted from this ADR.

Reviewer 4: ```suggestion
devices are exempted from this ADR.
```

## Consequences

We will no longer accept integrations that integrate with devices over GPIO as

> a core integration, those will have to be maintained as custom integrations,

Reviewer 4: ```suggestion
a core integration; those will have to be maintained as custom integrations,
```

## Decision

- We no longer accept integrations that integrate with devices over GPIO.
- We identify, deprecate and remove integrations that use GPIO.
- It will still be possible to have custom integrations that use GPIO.

## Consequences

We will no longer accept integrations that integrate with devices over GPIO as

> a core integration, those will have to be maintained as cutom integrations,

Reviewer: ```suggestion
a core integration, those will have to be maintained as custom integrations,
```
